### PR TITLE
feat(auth): Multi-tenancy support for Google Cloud Identity Platform

### PR DESCRIPTION
**Pull request by bojeil-google**
Merge all changes to master.

RELEASE NOTE: Added multi-tenancy support to the authentication service (Google Cloud Identity Platform project required). Tenant related APIs are exposed via [`tenantManager()`](https://firebase.google.com/docs/reference/admin/node/admin.auth.Auth#tenantManager) on the [`admin.auth`](https://firebase.google.com/docs/reference/admin/node/admin.auth) interface.
RELEASE NOTE: Added tenant management APIs `authForTenant()`, `getTenant()`, `listTenants()`, `deleteTenant()`, `createTenant()` and `updateTenant()` to the newly defined `TenantManager`.
RELEASE NOTE: Defined `TenantAwareAuth` interface retrieved via `TenantManager#authForTenant()` for managing users, configuring SAML/OIDC providers, generating email links for password reset, email verification, etc for specific tenants.

### CODE REVIEW

File: test/unit/auth/tenant.spec.ts [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+        expect(tenantWithoutAllowPasswordSignup.emailSignInConfig.passwordRequired).to.be.true;
+      }).not.to.throw();
+    });
+  });
+
+  describe('toJSON()', () => {
+    const serverRequestCopy: TenantServerResponse = deepCopy(serverRequest);
+    it('should return the expected object representation of a tenant', () => {
+      expect(new Tenant(serverRequestCopy).toJSON()).to.deep.equal({
+        tenantId: 'TENANT_ID',
```
Comment (wuyanna) [inline test/unit/auth/tenant.spec.ts]: TENANT-ID

File: test/unit/auth/tenant.spec.ts [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+        displayName: 'TENANT_DISPLAY_NAME',
```
Comment (wuyanna) [inline test/unit/auth/tenant.spec.ts]: TENANT-DISPLAY-NAME

File: test/unit/auth/user-import-builder.spec.ts [hunk 3]
```diff
@@ -66,6 +66,7 @@ describe('UserImportBuilder', () => {
         },
       ],
       customClaims: {admin: true},
+      tenantId: 'TENANT_ID',
```
Comment (wuyanna) [inline test/unit/auth/user-import-builder.spec.ts]: TENANT-ID

File: test/unit/auth/user-record.spec.ts [hunk 4]
```diff
@@ -626,6 +633,24 @@ describe('UserRecord', () => {
         (userRecord.providerData[0] as any).displayName = 'John Smith';
       }).to.throw(Error);
     });
+
+    it('should return undefined tenantId when not available', () => {
+      expect(userRecord.tenantId).to.be.undefined;
+    });
+
+    it('should return expected tenantId', () => {
+      const resp = deepCopy(getValidUserResponse('TENANT_ID'));
```
Comment (wuyanna) [inline test/unit/auth/user-record.spec.ts]: TENANT-ID

File: test/unit/auth/user-record.spec.ts [hunk 5]
```diff
@@ -626,6 +633,24 @@ describe('UserRecord', () => {
...
+      const tenantUserRecord = new UserRecord(resp);
+      expect(tenantUserRecord.tenantId).to.equal('TENANT_ID');
+    });
+
+    it('should throw when modifying readonly tenantId property', () => {
+      expect(() => {
+        const resp = deepCopy(getValidUserResponse('TENANT_ID'));
+        const tenantUserRecord = new UserRecord(resp);
+        (tenantUserRecord as any).tenantId = 'OTHER_TENANT_ID';
```
Comment (wuyanna) [inline test/unit/auth/user-record.spec.ts]: OTHER-TENANT-ID

File: test/unit/auth/tenant.spec.ts (hunk 1 above)
Comment (bojeil-google) [reply test/unit/auth/tenant.spec.ts]: done

File: test/unit/auth/tenant.spec.ts (hunk 2 above)
Comment (bojeil-google) [reply test/unit/auth/tenant.spec.ts]: done

File: test/unit/auth/user-import-builder.spec.ts (hunk 3 above)
Comment (bojeil-google) [reply test/unit/auth/user-import-builder.spec.ts]: done

File: test/unit/auth/user-record.spec.ts (hunk 4 above)
Comment (bojeil-google) [reply test/unit/auth/user-record.spec.ts]: done

File: test/unit/auth/user-record.spec.ts (hunk 5 above)
Comment (bojeil-google) [reply test/unit/auth/user-record.spec.ts]: done

**Pull request merged**